Add initial expressions clause (#35)
hlsl-tc57/tc57: spec/expressions.tex

\Ch{Expressions}{Expr}

\p This clause defines the formulations of expressions and the behavior of
operators when they are not overloaded. Operator overloading does not alter the
rules for operators defined by this standard.

\p An expression may also be an \textit{unevaluated operand} when it appears in
some contexts. An \textit{unevaluated operand} is a expression which is not
evaluated in the program.

\begin{note}
  The operand to \texttt{sizeof(...)} is a good example of an
  \textit{unevaluated operand}. In the code \texttt{sizeof(Foo())}, the call to
  \texttt{Foo()} is never evaluated in the program.
\end{note}

\p Whenever a \textit{glvalue} appears in an expression that expects a
\textit{prvalue}, a standard conversion sequence is applied based on the rules
in \ref{Conv}.

\Sec{Usual Arithmetic Conversions}{Expr.Conv}

\p Binary operators for arithmetic and enumeration type require that both
operands are of a common type. When the types do not match the \textit{usual

> arithmetic conversions} are applied to yield a common type. When \textit{usual

Reviewer: I will usually turn a blind eye to missing commas for dependent clauses, but this one ending in a verb with no object made it hard to read correctly. I first read not that the types were mismatching each other but that they were mismatching the `usual arithmetic conversions`, which clearly doesn't make sense, but that's how I parsed it. 
```suggestion
operands are of a common type. When the types do not match, the \textit{usual
arithmetic conversions} are applied to yield a common type. When \textit{usual
```

arithmetic conversions} are applied to vector operands they behave as
component-wise conversions (\ref{Conv.CWise}). The \textit{usual arithmetic
conversions} are:

\begin{itemize}

>   \item If either operand is of scoped enumeration type no conversion is

Reviewer: This one reads okay without the comma, but the rest of this list uses them.
```suggestion
  \item If either operand is of scoped enumeration type, no conversion is
```

    required.
    \item If one operand is a vector and the other operand is a scalar, the
    scalar is extended to a vector via a Splat conversion (\ref{Conv.VSplat}) to
    match the length of the vector.
    \item Otherwise, if both operands are vectors of different lengths, the
    vector of longer length is truncated to match the length of the shorter
    vector (\ref{Conv.VTrunc}).
  \end{itemize}
  \item If either operand is of type \texttt{double} or \texttt{vector<double,
  X>}, the other operator shall be converted to match.
  \item Otherwise, if either operand is of type \texttt{float} or \texttt{vector<float,
  X>}, the other operand shall be converted to match.
  \item Otherwise, if either operand is of type \texttt{half} or \texttt{vector<half, X>},
  the other operand shall be converted to match.
  \item Otherwise, integer promotions are performed on each scalar or vector
  operand following the appropriate scalar or component-wise conversion
  (\ref{Conv}).
  \begin{itemize}
    \item If both operands are scalar or vector elements of signed or unsigned
    types, the operand of lesser integer conversion rank shall be converted to

>     the type of the operand with greater rank.

Reviewer: Should this specify integer types?
```suggestion
    \item If both operands are scalar or vector elements of signed or unsigned
    integer types, the operand of lesser integer conversion rank shall be
    converted to the type of the operand with greater rank.
```

\begin{grammar}
  \define{unqualified-id}\br
  identifier\br
  operator-function-id\br
  conversion-function-id\br
  template-id\br
\end{grammar}

\SubSub{Qualified Identifiers}{Expr.Primary.ID.Qual}

\begin{grammar}
  \define{qualified-id}\br
  nested-name-specifier \opt{\keyword{template}} unqualified-id\br

  \define{nested-name-specifier}\br
  \terminal{::}\br
  type-name \terminal{::}\br
  namespace-name \terminal{::}\br
  nested-name-specifier identifier \terminal{::}\br
  nested-name-specifier \opt{\keyword{template}} simple-template-id \terminal{::}
\end{grammar}

\Sec{Postfix Expressions}{Expr.Post}

\begin{grammar}
  \define{postfix-expression}\br
  primary-expression\br
  % The [] characters on the two lines below should be inside \terminal, however
  % pandoc doesn't seem to like that.
  postfix-expression [ expression ]\br
  postfix-expression [ braced-init-list ]\br %
  postfix-expression \terminal{(} \opt{expression-list} \terminal{)}\br
  simple-type-specifier \terminal{(} \opt{expression-list} \terminal{)}\br
  typename-specifier \terminal{(} \opt{expression} \terminal{)}\br
  simple-type-specifier braced-init-list\br

>   typename-specifier braced-init-list\br

Reviewer: Lingering effects syntax prevents us from supporting these methods of initialization
```suggestion
```

Author: These syntaxes aren't supported by any HLSL compiler today (although admittedly they only fail in Clang due to a bug), so I'm happy to remove them.

That aside, I'm not particularly familiar with the effects syntax can you elaborate on how that impacts here? These are examples of these formulations in code:

```c++
template <typename T>
struct S {
    using type = T; 

    void fn() {
      auto v = typename S<T>::type{2}; // typename-specifier braced-init-list
    }
};

[numthreads(1,1,1)]
void main() {
    auto f = float{1}; // simple-type-specifier braced-init-list
}
```

Reviewer: To be honest, I'm not familiar with effects syntax either except that it involved a lot of {} and that DXC complains of ignoring deprecated syntax when I tried to use them for initializers : https://godbolt.org/z/7rzs1r5eM

In the example you give, I believe it's creating temporary float values, initializing them, then assigning them to the autos. In this case, DXC instead confuses the syntax for being casts and fails as well

The introduction of `auto` might make such formulations more interesting and we should make a note to consider it along with the other 202x features, which, happily, includes removal of effects syntax.

Author: I'm hesitant to incorporate any syntax that depends on braced-init-list before we fix braced initializers...

  postfix-expression \terminal{.} \opt{\terminal{template}} id-expression\br
  postfix-expression \terminal{.} vector-swizzle-component-sequence\br
  postfix-expression \terminal{.} matrix-swizzle-component-sequence\br
  postfix-expression \terminal{++}\br
  postfix-expression \terminal{--}
\end{grammar}

\Sec{Subscript}{Expr.Post.Subscript}

> % TODO: Language pending proposal in https://github.com/hlsl-tc57/tc57/pull/27.

Reviewer: As the PR has been merged, perhaps this should reference the proposal in the repo### 
```suggestion
% TODO: Language pending proposal in https://github.com/hlsl-tc57/tc57/blob/main/proposals/0006-restricted-unbound-arrays.md.
```

  \define{zero-index-value} \textnormal{one of}\br
  \terminal{0 1 2 3 }\br

  \define{matrix-one-indexed-swizzle-sequence}\br
  matrix-one-indexed-swizzle\br
  matrix-one-indexed-swizzle-sequence matrix-one-indexed-swizzle\br

  \define{matrix-one-indexed-swizzle}\br
  \terminal{\_} one-index-value one-index-value\br

  \define{one-index-value} \textnormal{one of}\br
  \terminal{1 2 3 4 }\br
\end{grammar}

\begin{itemize}
\item \textbf{.\_} (math-style): subsequent subscripts use \textbf{1-based}
 indexing for both row and column.
\item \textbf{.\_m} (memory-style): subsequent subscripts use \textbf{0-based}
 indexing for both row and column.
\end{itemize}

\p A \textit{matrix-swizzle-component-sequence} is a sequence of one but less
than or equal to four swizzle components of either the
\textit{matrix-zero-indexed-swizzle} or \textit{matrix-one-indexed-swizzle}
forms; the two forms may not be mixed in the same
\textit{matrix-swizzle-component-sequence}. The type of a swizzle expression is
a vector of the same element type as the postfix expression before the dot,
with a number of elements equal to the number of components in the

> \textit{matrix-swizzle-component-sequence}.

Reviewer: I just realized we might have technically expanded this when we introduced long vectors. Given how slightly odd it already is and how much odder it would be if you could chain up to 2048 swizzle components, I think I prefer we leave it at four.

is a member function name.

\p When a function is called, each parameter shall be initialized with its
corresponding argument. The order in which parameters are initialized is
unspecified.

\p If the function is a non-static member function the \texttt{this} argument
shall be initialized to a reference to the object of the call as if casted by an
explicit cast expression to an lvalue reference of the type that the function is
declared as a member of.

\p Parameters are either \textit{input parameters}, \textit{output parameters},
or \textit{input/output parameters} as denoted in the called function's
declaration (\ref{Decl.Function}). For all types of parameters the argument
expressions are evaluated before the function call occurs.

\p \textit{Input parameters} are passed by-value into a function. If an argument
to an \textit{input parameter} is of constant-sized array type, the array is
copied to a temporary and the temporary value is converted to an address via
array-to-pointer decay. If an argument is an unsized array type, the array
lvalue directly decays via array-to-pointer decay.

% TODO: The language above matches DXC's behavior but will need to be updated to

> % match the final specification from https://github.com/hlsl-tc57/tc57/pull/27.

Reviewer: ```suggestion
% match the final specification from https://github.com/hlsl-tc57/tc57/blob/main/proposals/0006-restricted-unbound-arrays.md.
```

from an argument to an \textit{input/output parameter} is initialized through
copy-initialization from the lvalue argument expression. Overload resolution
shall occur on argument initialization as if the expression \texttt{T Param =
Arg} were evaluated. In both cases, the cxvalue shall have the type of the
parameter and the argument can be converted to that type through implicit or
explicit conversion.

\p If an argument to an \textit{output} or \textit{input/output parameter} is a
constant sized array, the array is copied to a temporary cxvalue following the
same rules for any other data type. If an argument to an \textit{output} or
\textit{input/output parameter} is an unsized array type, the array lvalue
directly decays via array-to-pointer decay. An argument of a constant sized
array of type \texttt{T[N]} can be converted to a cxvalue of an unsized array
of type \texttt{T[]} through array to pointer decay. An unsized array of type
\texttt{T[]}, cannot be implicitly converted to a a constant sized array of type
\texttt{T[N]}.

% TODO: The language above matches DXC's behavior but will need to be updated to

> % match the final specification from https://github.com/hlsl-tc57/tc57/pull/27.

Reviewer: ```suggestion
% match the final specification from https://github.com/hlsl-tc57/tc57/blob/main/proposals/0006-restricted-unbound-arrays.md.
```

is passed by-address to the caller.

\begin{note}
  If the argument to an \textit{output} or \textit{input/output parameter} is
  the result of a side-effecting expression, the side-effecting expression is
  evaluated only once, and the result is stored in a temporary cxvalue which is
  passed by-address to the caller. The temporary cxvalue is then assigned back
  to the argument expression after.
\end{note}

\p When a function is called, any parameter of object type must have completely
defined type, and any parameter of array of object type must have completely
defined element type. The lifetime of a parameter ends on return of the function
in which it is defined. Initialization and destruction of each
parameter occurs within the context of the calling function.

\p The value of a function call is the value returned by the called function.

\p A function call is an lvalue if the result type is an lvalue reference type;
otherwise it is a prvalue.

\begin{note}
  Functions that return lvalue references are unspellable in HLSL, but the
  language requires them for some built-in functions. For example, the

>   \texttt{operator[]} overloads for writable resource types.

Reviewer: From a language-perspective, that's a different postfix expression. I don't know of any built-in functions that do this

Author: It isn't strictly a different postfix expression because it can be written of the form:

```hlsl
RWBuffer<float> In;

[numthreads(1,1,1)]
void main() {
    In.operator [](1);
}
```

I believe the other occurrence is the `Get` functions on the work graphs types.

Reviewer: The first is a pretty strange thing to do, but the `Get`s are the designated way to do things. Thanks for the insight
